### ISSUE TITLE
Primitive Schema Should Support More Data Types

### ISSUE THREAD

**Issue by shaggygi** [Azure/IoTPlugandPlay#8]
The following data types should be added to the [Primitive Schema list/table](https://github.com/Azure/IoTPlugandPlay/tree/master/DTDL#primitive-schemas).

| Digital twin primitive schema | Description |
| --- | --- |
| byte | An unsigned 8-bit integer. |
| short | A signed 2-byte integer. |
| uint | An unsigned 4-byte integer. |
| ulong | An unsigned 8-byte integer. |
| ushort | An unsigned 2-byte integer. |

**Comment by TanmayBhagwat** [comment]
Thanks for feedback Greg! We really appreciate it. 

Could you give us examples of scenarios where you'll want to use those data type?

**Comment by shaggygi** [comment]
@Tanmaiye Yes, especially when it comes to IoT devices (memory registers and such).  They are usually in sizes of 1/2/4/8/etc bytes.  Therefore, there will be cases when unsigned values need to be represented instead of being required to convert between unsigned & signed.

For example, you might have an 16-bit register where each bit represents the input/output direction for an I/O expander device.  Let's say 1 = input and 0 = output and we have the top byte all inputs and lower byte all outputs.  So the value would be 0b11111111_00000000.  Having the ability to use unsigned data types would be simple compared to signed values more conversions would be required.

You see lots of this work going on we're doing on the [.NET IoT repo](https://github.com/dotnet/iot).

Hope this helps.

**Comment by shaggygi** [comment]
How would you represent **bits** within other primitive data types (byte, integer, ushort, etc.)?  Understanding, you currently can have a **boolean** individually like below.

```json
{
          "@id": "http://azureiot.com/interfaces/deviceInformation/1/statusled",
          "@type": "Property",
          "comment": "LED that represents the current status of IoT device.",
          "displayName": "Status LED",
          "name": "statusLed",
          "schema": "boolean"
}
```

Related to System.Device.GPIO APIs, the above would be helpful as the property could be converted to control the hardware LED.  See [blinking code example here](https://github.com/dotnet/iot/blob/master/samples/led-blink/README.md#code).

However, there are many cases where byte(s) are read/written as a whole, but are defined by multiple bits that represent hardware flags, interrupts, I/O, etc..  Therefore, we need a way to define data types down to the bit-level.  For example, lets say there is a byte data type (e.g. SPI memory register) where bit 0 represents an interrupt, bit 3 represents the state of an input, bit 6 represents an error LED and bit 7 represents a status LED.

**Disclaimer**: I'm only thinking out loud in this example to try and explain for collaboration.

```json
{
          "@id": "http://azureiot.com/interfaces/deviceInformation/1/controlRegister",
          "@type": "Property",
          "comment": "Register that represents various I/O.",
          "displayName": "Control Register",
          "name": "controlRegister",
          "schema": "byte",
          "bits": [
            {
                    "@id": "http://azureiot.com/interfaces/deviceInformation/1/controlRegister/input1Interrupt",
                    "@type": "Property",
                    "comment": "Interrupt flag for input 1.",
                    "displayName": "Input 1 Interrupt",
                    "name": "input1Interrupt",
                    "schema": "boolean",
                    "bit": "0",
        },
        {
                    "@id": "http://azureiot.com/interfaces/deviceInformation/1/controlRegister/input1",
                    "@type": "Property",
                    "comment": "Input 1 located on hardware.",
                    "displayName": "Input 1",
                    "name": "input1",
                    "schema": "boolean",
                    "bit": "3",
        },
        {
                    "@id": "http://azureiot.com/interfaces/deviceInformation/1/controlRegister/errorLed",
                    "@type": "Property",
                    "comment": "LED that represents an error has occurred..",
                    "displayName": "Error LED",
                    "name": "errorLed",
                    "schema": "boolean",
                    "bit": "6",
        },
        {
                    "@id": "http://azureiot.com/interfaces/deviceInformation/1/controlRegister/statusLed",
                    "@type": "Property",
                    "comment": "LED that represents the current status of IoT device.",
                    "displayName": "Status LED",
                    "name": "statusLed",
                    "schema": "boolean",
                    "bit": "7",
        },
          ]
}
```

**NOTE**: This example only focused on individual bits, but could extend this even further for enums and multiple bits with a data type to represent (but that's for another day... 😄)

**Comment by shaggygi** [comment]
How would you configure for choices?  Enums are an option, but what if you only wanted to limit a property to only 2 selections within the entire enum values?  Or what if you need to choose between properties that are of different types?  Meaning, I need to select a Boolean, Integer or complex object.
